learning_resources_search/tasks.py

```diff
@@ -108,6 +115,111 @@ def upsert_learning_resource(learning_resource_id):
+        if key in group_keys and val:
+            if key == "department":
+                return LearningResourceDepartment.objects.get(department_id=val[0]).name
+            elif key == "offered_by":
+                return LearningResourceOfferor.objects.get(code=val[0]).name
+            return val[0]
+    return None
+
+
+def _infer_search_url(percolate_query):
+    """
+    Infer the search URL for the percolate query
+    """
+    original_query = OrderedDict(percolate_query.original_query)
+    query_string_params = {k: v for k, v in original_query.items() if v}
+    query_string = urlencode(query_string_params, doseq=True)
+    return f"{settings.SITE_BASE_URL}/search?{query_string}"
+
+
+def _group_percolated_rows(rows):
+    def key_func(x):
+        return (x["user_id"], x["group"])
+
+    grouped_data = {}
+    for key, group in itertools.groupby(rows, key_func):
+        context = list(group)
+        user_id = key[0]
+        group_name = key[1]
+        if key[0] not in grouped_data:
+            grouped_data[user_id] = {group_name: []}
+        if group_name not in grouped_data[user_id]:
+            grouped_data[user_id][group_name] = []
+        for ctx in context:
+            if ctx["user_id"] == user_id:
+                grouped_data[user_id][group_name].append(ctx)
+    return grouped_data
+
+
+def _get_percolated_rows(resources, subscription_type):
+    """
+    Get percolated rows for a list of learning resources and subscription type
+    """
+    rows = []
+    all_users = set()
+    # percolate each new learning resource to get matching queries
+    for resource in resources:
+        percolated = percolate_matches_for_document(resource.id).filter(
+            source_type=subscription_type
+        )
+        if percolated.count() > 0:
+            percolated_users = set(percolated.values_list("users", flat=True))
+            all_users.update(percolated_users)
+            query = percolated.first()
+            rows.extend(
+                [
+                    {
+                        "resource": resource,
+                        "user_id": user,
+                        "query": query,
+                        "group": _infer_percolate_group(query),
+                        "search_url": _infer_search_url(query),
+                    }
+                    for user in percolated_users
+                ]
+            )
+
+    return rows
+
+
+@app.task(autoretry_for=(RetryError,), retry_backoff=True, rate_limit="600/m")
+def send_subscription_emails(subscription_type, period="daily"):
+    """
+    Send subscription emails by percolating matched documents
+    """
+    delta = datetime.timedelta(days=1)
+    if period == "weekly":
+        delta = datetime.timedelta(days=7)
+    since = now_in_utc() - delta
+    new_learning_resources = LearningResource.objects.filter(
+        published=True, created_on__gt=since
+    )
+    rows = _get_percolated_rows(new_learning_resources, subscription_type)
```

Comment:
This variable `rows` always ends up as an empty list for me, so no emails are sent.  I subscribed to 3 channels (1 offeror, 1 topic, 1 department).  Added some logging and all published courses in my database are included in `new_learning_resources` (because I reingested them all from scratch).

I managed to get an email by doing the following, and it looks good:

```python
query = PercolateQuery.objects.first()
lr = LearningResource.objects.first()
rows = [{"resource": lr, "user_id": 1, "query": query, "group": _infer_percolate_group(query),  "search_url": _infer_search_url(query)}]
template_data = _group_percolated_rows(rows)
for user_id in template_data:
    user = User.objects.get(id=user_id)
    total_count = sum(
        [len(template_data[user_id][group]) for group in template_data[user_id]]
    )
    send_template_email(
        [user.email],
        f"{settings.MITOPEN_TITLE} New Learning Resources for You",
        "email/subscribed_channel_digest.html",
        context={"documents": template_data[user_id], "total_count": total_count},
    )
```

Reply:
This might be because the percolate queries hadnt indexed yet. If you do a re-index it should populate.

Reply:
PS the link in the email to view the search results contained the parameter `&endpoint=learning_resource` which can probably be omitted?

Reply:
Getting rows now, the lack of them was due to my frontend not updating properly until I deleted all my `node_modules` and `build` folders.

The "view more" link still contains an `endpoint=learning_resource` parameter though, which is ignored: http://open.odl.local:8063/search?endpoint=learning_resource&department=8

Reply:
added code to pop the endpoint parameter.